src/Mono.Android/Android.Runtime/JNIEnv.cs

```diff
@@ -249,15 +236,56 @@ static void ManualJavaObjectDispose (Java.Lang.Object obj)
 			GC.SuppressFinalize (obj);
 		}
 
+		static Action<Exception> mono_unhandled_exception;
+		static Action<AppDomain, UnhandledExceptionEventArgs> AppDomain_DoUnhandledException;
+
+		static void InitializeUncaughtExceptionHandling ()
+		{
+			if (mono_unhandled_exception == null) {
+				var mono_UnhandledException = typeof (System.Diagnostics.Debugger)
+					.GetMethod ("Mono_UnhandledException", BindingFlags.NonPublic | BindingFlags.Static);
+				mono_unhandled_exception = (Action<Exception>) Delegate.CreateDelegate (typeof(Action<Exception>), mono_UnhandledException);
+			}
+
+			if (AppDomain_DoUnhandledException == null) {
+				var ad_due = typeof (AppDomain)
+					.GetMethod ("DoUnhandledException",
+					            bindingAttr:  BindingFlags.NonPublic | BindingFlags.Instance,
+					            binder:       null,
+					            types:        new []{typeof (UnhandledExceptionEventArgs)},
+					            modifiers:    null);
+				if (ad_due != null) {
+					AppDomain_DoUnhandledException  = (Action<AppDomain, UnhandledExceptionEventArgs>) Delegate.CreateDelegate (
+						typeof (Action<AppDomain, UnhandledExceptionEventArgs>), ad_due);
+				}
+			}
+		}
+
 		internal static void PropagateUncaughtException (IntPtr env, IntPtr javaThreadPtr, IntPtr javaExceptionPtr)
 		{
-			if (defaultUncaughtExceptionHandler == null)
+			if (!PropagateExceptions)
 				return;
 
-			var javaThread = JavaObject.GetObject<Java.Lang.Thread> (env, javaThreadPtr, JniHandleOwnership.DoNotTransfer);
+			try {
+				InitializeUncaughtExceptionHandling ();
+			} catch (Exception e) {
+				Android.Runtime.AndroidEnvironment.FailFast ($"Unable to initialize UncaughtExceptionHandler. Nested exception caught: {e}");
+			}
+
 			var javaException = JavaObject.GetObject<Java.Lang.Throwable> (env, javaExceptionPtr, JniHandleOwnership.DoNotTransfer);
 
-			defaultUncaughtExceptionHandler.UncaughtException (javaThread, javaException);
+			mono_unhandled_exception (javaException);
+			if (AppDomain_DoUnhandledException == null)
+				return;
+
+			try {
+				var jltp = javaException as JavaProxyThrowable;
+				Exception innerException = jltp?.InnerException;
+				var args  = new UnhandledExceptionEventArgs (innerException ?? javaException, isTerminating: true);
+				AppDomain_DoUnhandledException (AppDomain.CurrentDomain, args);
+			} catch (Exception e) {
+				Logger.Log (LogLevel.Error, "monodroid", "Exception thrown while raising AppDomain.UnhandledException event: " + e.ToString ());
+			}
```

Comment:
Since `mono-2019-12`, I think we can do this now: https://github.com/xamarin/xamarin-android/pull/3991

We have `InternalsVisibileTo` and we should be able to call either of these directly without reflection:
```csharp
System.Diagnostics.Debugger.Mono_UnhandledException (ex);
AppDomain.CurrentDomain.DoUnhandledException (args);
```

https://github.com/mono/mono/pull/18040

Reply:
Good catch, thanks!